sdk/python/kfp/_client.py

```diff
@@ -302,6 +304,37 @@ def _refresh_api_client_token(self):
     new_token = get_gcp_access_token()
     self._existing_config.api_key['authorization'] = new_token
 
+  def _get_config_with_default_credentials(self, config):
+    """Apply default credentials to the configuration object.
+
+    This method accepts a Configuration object and extends it with some default
+    credentials interface.
+    """
+    # XXX: The default credentials are audience-based service account tokens
+    # projected by the kubelet (ServiceAccountTokenVolumeCredentials). As we
+    # implement more and more credentials, we can have some heuristic and
+    # choose from a number of options.
+    # See https://github.com/kubeflow/pipelines/pull/5287#issuecomment-805654121
+    credentials = ServiceAccountTokenVolumeCredentials()
+
+    try:
+        token = credentials.get_token()
+    except OSError as e:
```

Comment:
credentials should be the interface, if there are ignorable errors, they should handle that by themselves.

We cannot catch exceptions thrown when refreshing credentials

Reply:
That's a good thought. I moved the error handling back to the credentials, and it returns `None`.
This part, I'm not very fond of it. Now this method checks `if not credentials.get_token()` to abort setting default credentials. If you have some better idea, feel free to suggest

Reply:
@elikatsis I'd prefer that we only call the `credentials.refresh_api_key` hook and avoid this `get_token`.

For example, if a auth implementation overrides refresh_api_key method and adds token to cookies, what should `get_token` do? Therefore, I think we should avoid `get_token` altogether, it's an implementation detail.

Reply:
@Bobgy, I see what you mean. I can't tell if this solution works with cookies, because I don't know what it requires.

If you check the description of the code & the links I have included in https://github.com/kubeflow/pipelines/issues/5138#issuecomment-797003710, search for `get_api_key_with_prefix`, all of this is tied to headers and especially the `authorization` header.

Thinking while writing, `refresh_api_key_hook` can modify any field of the `Configuration` as each implementation desires. So this can help with the cookie settings, as long as you also configure some `authorization` header or something?

I'll look into it a bit more and will comment back. I believe we can do as you propose, but I want to make sure it's solid.

Reply:
@Bobgy I moved the abstraction to the `refresh_api_key_hook` method instead of `get_token`. Please tell me what you think.